[FLINK-13375][table-api] Improve config names in ExecutionConfigOptions and OptimizerConfigOptions

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request moves `ExecutionConfigOptions` and `OptimizerConfigOptions` into API module. We actually only copy them to API module for reviewing. Will refactor blink planner to use the API options in a following PR.

## Brief change log

- moves `ExecutionConfigOptions` and `OptimizerConfigOptions` into API module. 


## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

wuchong commented:
cc @twalthr @aljoscha @dawidwys . We can review and discuss config option names in this pull request.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -0,0 +1,202 @@
+	//  Resource Options
+	// ------------------------------------------------------------------------
+
+	public static final ConfigOption<Integer> SQL_RESOURCE_DEFAULT_PARALLELISM =
```
All options in this class should be prefixed with `table.exec.`. If not, move them into a separate class?

wuchong replied:
+1 to `.exec.resource`

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -0,0 +1,202 @@
+					.withDescription("Sets the externalBuffer memory size that is used in sortMergeJoin and overWindow.");
+
+	public static final ConfigOption<Integer> SQL_RESOURCE_HASH_AGG_TABLE_MEM =
+			key("sql.resource.hash-agg.table.memory.mb")
```
The `.mb` syntax is deprecated. See `org.apache.flink.configuration.TaskManagerOptions#TASK_MANAGER_HEAP_MEMORY` vs `org.apache.flink.configuration.TaskManagerOptions#TASK_MANAGER_HEAP_MEMORY_MB`

wuchong commented:
Hi @twalthr , I agree with the hierarchical naming. 

Regarding to why choosing the `sql.` prefix, there are 3 reasons:
1. The `sql.` prefix is more widely used in Hadoop ecosystem, for example, `spark.sql.xxx` instead of `spark.dataframe.xxx`. We also want to promote SQL as the first citizen in Table API / SQL.
2. The `table.` prefix  may confuse users what "table" does it mean? The table module or `Table` in Table API or table in database? 
3. We have some options contains `table` in the key, for example: `sql.resource.hash-agg.table.memory.mb`, `sql.optimizer.reuse.table-source.enabled`. If we use `table.` prefix, it may conflict with the "table" in the option name. 

What do you think? @twalthr  @godfreyhe

twalthr commented:
In my opinion we should distinguish between the SQL language and the table ecosystem.

1. In most presentations, documentation, and in discussions we refer to the "table ecosystem" or "Table API" than "SQL API".
2. This is also represented in the Maven module names. `flink-sql-client` and `flink-sql-parser` are around SQL. `flink-table-XXX` are broader and cover working with the concept of "tables" and relational algebra (independent of SQL).
3. The Table API will receive a lot of new features such as machine learning via `flatMap, flatAggregate` etc. in the future. This has nothing to do with SQL and would confuse if runtime options are named like this.

Any other opinions @dawidwys @aljoscha?

dawidwys commented:
Generally speaking I also prefer the `table` prefix. I agree with @twalthr that in Flink's naming conventions the module that works with the relational algebra was usually referred as a "table ecosystem" or "Table API", of which the SQL language is part.

Some additional comments regarding the @wuchong's points
Ad.1 I am not aware of any decisions that we want to make the SQL string representation as the go-to interface. I think what we want is for the relational Table API (both sql as well as the java/scala counterparts) to be the first class citizen among different APIs like DataStream/DataSet.
Ad.2 To be honest I could revert the point with just substituting `table.` with `sql.` ;) Does it mean the sql in `.sqlUpdate`? Does it mean the `Table API` or just the sql subpart?
Ad.3 I don't think this is a problem. Moreover I think some of the mentioned options actually abuse the term `table`. E.g. in `sql.resource.hash-agg.table.memory.mb` I think it actually refers to managed memory of the hash aggregate operator. What does the `table` there means? 
On the other hand in the option `sql.optimizer.reuse.table-source.enabled` there is no table, only `table-source`. Theoretically we could even think about dropping the `table-` from `table-source` in that option.

wuchong commented:
Thanks for your feedbacks @twalthr @dawidwys , I think advanced feature of Table API `flatMap`, `flatAggregate` is a good reason to go with `table.` prefix. 

So I'm fine with the `table.` prefix. Do you have some thoughts on this @godfreyhe ?

godfreyhe commented:
`sql` is more popular than `table` for blink users before, while  for community users `table` is more popular. so I think `table` is better than `sql`.

wuchong commented:
Hi @twalthr , @dawidwys , I have updated the pull request.

I also update some option names according to the hierarchical naming convention and examples Timo pointed out. 

But I didn't adopt `.enabled` suffix changes, for example `table.exec.sort.async-merge.enabled` -> `table.exec.sort.async-merge-enabled`. I still keep it `table.exec.sort.async-merge.enabled`. Because `async-merge` is an individual word and may have sub-entries in the future, e.g. threads-num. The same to the other `enabled` options: `table.optimizer.reuse.sub-plan.enabled`, `table.optimizer.reuse.source.enabled`, `table.optimizer.predicate-pushdown.enabled`, `table.optimizer.join-reorder.enabled`. IMO, it will be more readable if separate `enabled` with the word before.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -83,97 +83,97 @@
+	public static final ConfigOption<String> TABLE_EXEC_RESOURCE_HASH_JOIN_TABLE_MEM =
+			key("table.exec.resource.hash-join.memory")
+					.defaultValue("128 mb")
+					.withDescription("Sets the managed memory for hash join operator. It defines the lower limit.");
```
nit: could you also update the typos in other descriptions such as `externalBuffer` -> `external buffer`, `sortMergeJoin` -> `sort-merge join` etc.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -83,97 +83,97 @@
-	public static final ConfigOption<Integer> SQL_EXEC_LOOKUP_ASYNC_BUFFER_CAPACITY =
-			key("sql.exec.lookup.async.buffer-capacity")
+	public static final ConfigOption<Integer> TABLE_EXEC_LOOKUP_ASYNC_BUFFER_CAPACITY =
+			key("table.exec.lookup.async.buffer-capacity")
```
`table.exec.lookup.async-buffer-capacity` or just `table.exec.lookup.buffer-capacity`?

wuchong replied:
Currently, we have two implementation for lookup a dimension table, one is synchronically, the other is asynchronously. So I think `async` is needed here. I'm not sure whether we will have any options for synchronically lookup. So I would still +1 to `table.exec.async.xxx`.

wuchong replied:
Or how about `table.exec.async-lookup.buffer-capacity` ?

twalthr replied:
Sounds good.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -83,97 +83,97 @@
-	public static final ConfigOption<String> SQL_EXEC_LOOKUP_ASYNC_TIMEOUT =
-			key("sql.exec.lookup.async.timeout")
+	public static final ConfigOption<String> TABLE_EXEC_LOOKUP_ASYNC_TIMEOUT =
+			key("table.exec.lookup.async.timeout")
```
`table.exec.lookup.async-timeout` or just `table.exec.lookup.timeout`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -83,97 +83,97 @@
-	public static final ConfigOption<String> SQL_EXEC_STATE_TTL =
-			key("sql.exec.state.ttl")
+	public static final ConfigOption<String> TABLE_EXEC_STATE_TTL =
+			key("table.exec.state.ttl")
```
`table.exec.state-ttl`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -30,8 +30,8 @@
-	public static final ConfigOption<String> SQL_OPTIMIZER_AGG_PHASE_STRATEGY =
-			key("sql.optimizer.agg-phase.strategy")
+	public static final ConfigOption<String> TABLE_OPTIMIZER_AGG_PHASE_STRATEGY =
+			key("table.optimizer.agg-phase.strategy")
```
`table.optimizer.agg-phase-strategy`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -56,33 +56,33 @@
-	public static final ConfigOption<Boolean> SQL_OPTIMIZER_REUSE_SUB_PLAN_ENABLED =
-			key("sql.optimizer.reuse.sub-plan.enabled")
+	public static final ConfigOption<Boolean> TABLE_OPTIMIZER_REUSE_SUB_PLAN_ENABLED =
+			key("table.optimizer.reuse.sub-plan.enabled")
```
`table.optimizer.reuse-sub-plan-enabled`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -56,33 +56,33 @@
-	public static final ConfigOption<Boolean> SQL_OPTIMIZER_REUSE_TABLE_SOURCE_ENABLED =
-			key("sql.optimizer.reuse.table-source.enabled")
+	public static final ConfigOption<Boolean> TABLE_OPTIMIZER_REUSE_SOURCE_ENABLED =
+			key("table.optimizer.reuse.source.enabled")
```
`table.optimizer.reuse-source-enabled`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -56,33 +56,33 @@
-	public static final ConfigOption<Boolean> SQL_OPTIMIZER_PREDICATE_PUSHDOWN_ENABLED =
-			key("sql.optimizer.predicate-pushdown.enabled")
+	public static final ConfigOption<Boolean> TABLE_OPTIMIZER_PREDICATE_PUSHDOWN_ENABLED =
+			key("table.optimizer.predicate-pushdown.enabled")
```
`table.optimizer.predicate-pushdown-enabled`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -56,33 +56,33 @@
-	public static final ConfigOption<Boolean> SQL_OPTIMIZER_JOIN_REORDER_ENABLED =
-			key("sql.optimizer.join-reorder.enabled")
+	public static final ConfigOption<Boolean> TABLE_OPTIMIZER_JOIN_REORDER_ENABLED =
+			key("table.optimizer.join-reorder.enabled")
```
`table.optimizer.join-reorder-enabled`?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -83,97 +83,97 @@
-	public static final ConfigOption<Integer> SQL_RESOURCE_SOURCE_PARALLELISM =
-			key("sql.resource.source.parallelism")
+	public static final ConfigOption<Integer> TABLE_EXEC_RESOURCE_SOURCE_PARALLELISM =
+			key("table.exec.resource.source.parallelism")
```
`source-parallelism`, `sink-parallelism`, `external-buffer-memory`, `hash-agg-memory` etc.? A hierarchy with one child is not necessary.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -56,33 +56,33 @@
+	public static final ConfigOption<Boolean> TABLE_OPTIMIZER_JOIN_REORDER_ENABLED =
+			key("table.optimizer.join-reorder.enabled")
 					.defaultValue(false)
 					.withDescription("Enables join reorder in optimizer cbo. Default is disabled.");
```
what is "cbo"?

godfreyhe replied:
cost based optimizer (e.g. VolcanoPlanner in Calcite)
another concept is "rbo", means rule based optimizer (e.g. HepPlanner in Calcite)

wuchong replied:
@godfreyhe , is it only enabled in CBO? I mean, users shouldn't care about what optimizer is used in Flink.

wuchong commented:
Thanks @twalthr for the detailed review. 

Regarding to the `.enabled` options, I left some thoughts above. I'm afraid we will have more sub-entries for those `.enabled` options. Take join-reorder as an example, spark has `spark.sql.cbo.joinReorder.enabled`, `spark.sql.cbo.joinReorder.dp.threshold`, `spark.sql.cbo.joinReorder.card.weight`. That's why I want to keep `.enabled`. And I think a hierarchy with one child can be valid as long as it is a reasonable word and may have sub-entries in the future.

What do you think/

twalthr commented:
I think if we add a hierarchy in advance then it should be aligned by operator and not by functionality. 

For example, instead of `table.optimizer.join-reorder.enabled` how about `table.optimizer.join.reorder-enabled`? Such that all join related config is grouped. Or instead of `table.optimizer.predicate-pushdown-enabled` how about `table.optimizer.source.predicate-pushdown-enabled`? 

We should definitely avoid having more then 4 levels because it becomes hart to read.

wuchong commented:
FLINK-13266 is just merged. We will keep the conflicts to make the reviewing easier. Once we reach a consensus on each config names. I will rebase master and adapt blink planner to the new options.

wuchong commented:
Hi @twalthr , I agree we should keep config names short. However, IMO, 4 levels is too strict, because in fact, we uses only have 2 levels to describe a config. There are some configs are difficult to squash to 4 levels. For example: `table.optimizer.distinct-agg.split.bucket-num` and `table.optimizer.distinct-agg.split.enabled`. I think maybe we can relax the restrictions to 5 levels?

Regarding to the aligned by operator vs functionality, I would like to align by functionality. Because some options are not relative to a specific operator, e.g `table.optimizer.reuse.sub-plan.enabled`. And some functionality words are well-known, e.g. `hash-agg`, `join-reorder`.

Regarding to the `.enabled`, my concern is 
1. it would be better to align all `enabled` options in the same style
2. using `.enabled` is more flexible for introducing new options under the same parent.

> we don't need a hierarchy because they just consist of a single child.

I think it's fine to have a hierarchy even if it consist of a single child. Because we can't enumerate all the configs at one blow. We should leave some space for the future. And the [Flink convention](https://docs.google.com/document/d/1JGRVK6udBFa8xbyv7ipPQiLEhxebA64f8KKcaAH0UPY/edit#) doesn't restrict this, right?

What do you think? @twalthr @godfreyhe

godfreyhe commented:
I think readability and extendibility of a config name is more important than strict level. From this point of view, I also tend to align by functionality instead of operator. e.g. `lookup` has `async` and `sync` mode, I think `table.exec.async-lookup.XXX` and `table.exec.sync-lookup.XXX` is more clear than `table.exec.lookup.async-XXX` and `table.exec.lookup.sync-XXX`.

about `enabled`,  I think the readability is almost the same about `table.exec.sort.async-merge.enabled` and `table.exec.sort.async-merge-enabled`, however the extendibility of the first one is better than second one.

twalthr commented:
5 Levels for special cases like `table.optimizer.distinct-agg.split.bucket-num` is fine with me.

However, I think if we introduce new options under a common key that needs a hierarchy, we could still deprecate the old flat key and introduce the hierarchy. Introducing hierarchy "just in case" should be avoided.

If options are not relative to a specific operator, they should become top-level properties: `table.optimizer.reuse-sub-plan-enabled` instead of `table.optimizer.reuse.sub-plan.enabled`.

I know this topic is typical bikeshedding so I don't want to block too much here. But it would be great if you could go through the list one more time and check whether a hierarchy makes sense. Thanks.

wuchong commented:
Thanks for the feedbacks. I have update the config names again. 
Welcome to have another look at it @twalthr @godfreyhe @dawidwys .

godfreyhe commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -0,0 +1,202 @@
+					.defaultValue(false)
+					.withDescription("Specifies whether to enable MiniBatch optimization. " +
+							"MiniBatch is an optimization to buffer input records to reduce state access. " +
+							"This is disabled by default. To enable this, users should set this config to true.");
```
add description: "NOTE: If this is true, TABLE_EXEC_MINIBATCH_ALLOW_LATENCY.key() and TABLE_EXEC_MINIBATCH_SIZE.key() must be greater than zero."

godfreyhe commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -0,0 +1,202 @@
+	// ------------------------------------------------------------------------
+	//  Other Exec Options
+	// ------------------------------------------------------------------------
+	public static final ConfigOption<String> TABLE_EXEC_DISABLED_OPERATORS =
```
list current supported operator names ?

wuchong commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -0,0 +1,88 @@
+			key("table.optimizer.distinct-agg.split.enabled")
+					.defaultValue(false)
+					.withDescription("Tells the optimizer whether to split distinct aggregation " +
+							"(e.g. COUNT(DISTINCT col), SUM(DISTINCT col)) into two level. " +
```
This is on purpose. The two level is different two phase in `table.optimizer.agg-phase-strategy`. They are orthogonal.

wuchong commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -0,0 +1,88 @@
+			key("table.optimizer.distinct-agg.split.bucket-num")
+					.defaultValue(1024)
+					.withDescription("Configure the number of buckets when splitting distinct aggregation. " +
+							"The number is used in the first level aggregation to calculate a bucket key " +
```
The same as above.

wuchong commented:
Thanks for the reviewing @godfreyhe . I have updated the PR.
@twalthr do you want to go through it again?

wuchong commented:
Thanks @twalthr . I will adapt blink planner to the new config names then.

wuchong commented:
Hi @JingsongLi @godfreyhe , I have updated the pull request to adapt blink planner to the new config names. In order for easy reviewing, I split it into 2 commits:
1. the first one is just renaming config field names. It touches many files but all is renaming.
2. the second one is renaming config key names, improving description, changing `Integer` option -> `String` option (memory size).

Feel free to leave comments. :)

godfreyhe commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:
```diff
@@ -128,80 +123,71 @@
+			.withDescription("Specifies whether to enable MiniBatch optimization. " +
+				"MiniBatch is an optimization to buffer input records to reduce state access. " +
+				"This is disabled by default. To enable this, users should set this config to true. " +
+				"NOTE: If mini-batch is enabled, 'table.exec.mini-batch.allow-latency' and " +
```
use `TABLE_EXEC_MINIBATCH_ALLOW_LATENCY.key()` instead of `table.exec.mini-batch.allow-latency` ?  and `table.exec.mini-batch.size`

wuchong replied:
We can't do that. Java will complain "Illegal forward reference" for that.

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/BatchExecSortRule.scala:
```diff
@@ -81,8 +81,8 @@ object BatchExecSortRule {
-  val SQL_EXEC_SORT_RANGE_ENABLED: ConfigOption[JBoolean] =
-  key("sql.exec.sort.range.enabled")
+  val TABLE_EXEC_SORT_RANGE_ENABLED: ConfigOption[JBoolean] =
+  key("table.exec.sort.range.enabled")
```
-> `table.exec.range-sort.enabled` ?

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecSort.scala:
```diff
@@ -146,8 +146,8 @@ object StreamExecSort {
-  val SQL_EXEC_SORT_NON_TEMPORAL_ENABLED: ConfigOption[JBoolean] =
-  key("sql.exec.sort.non-temporal.enabled")
+  val TABLE_EXEC_SORT_NON_TEMPORAL_ENABLED: ConfigOption[JBoolean] =
+  key("table.exec.sort.non-temporal-enabled")
```
-> `table.exec.non-temporal-sort.enabled` ?

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecRank.scala:
```diff
@@ -246,8 +246,8 @@ object StreamExecRank {
-  val SQL_EXEC_TOPN_CACHE_SIZE: ConfigOption[JLong] =
-  key("sql.exec.topn.cache.size")
+  val TABLE_EXEC_TOPN_CACHE_SIZE: ConfigOption[JLong] =
+  key("table.exec.topn.cache.size")
```
-> `table.exec.topn.cache-size` ?

wuchong commented:
Thanks for the reviewing @godfreyhe , I will modify them when merging because they are tiny changes.
